File: 0-high-concurrency-inventory-management/1-java/src/main/java/com/starci/inventory/InventoryService.java
package com.starci.inventory;

import java.util.List;

import jakarta.annotation.PostConstruct;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.data.redis.core.script.RedisScript;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Service
public class InventoryService {

    private static final String SEEDED_SKU = "IPHONE15";

    // Lua runs single-threaded inside Redis: check-and-decrement is one indivisible step.
    private static final RedisScript<Long> DECREMENT_LUA = RedisScript.of("""
            local stock = tonumber(redis.call('GET', KEYS[1]))
            if stock == nil then return -1 end
            if stock < tonumber(ARGV[1]) then return -2 end
            return redis.call('DECRBY', KEYS[1], ARGV[1])
            """, Long.class);

    private final StringRedisTemplate redis;
    private final InventoryRepository inventories;
    private final InventoryLedgerRepository ledgers;

    public InventoryService(StringRedisTemplate redis,
                            InventoryRepository inventories,
                            InventoryLedgerRepository ledgers) {
        this.redis = redis;
        this.inventories = inventories;
        this.ledgers = ledgers;
    }

    /** Warm the hot Redis counter from the authoritative DB stock on startup. */
    @PostConstruct
    public void warmCounter() {
        inventories.findById(SEEDED_SKU)
                .ifPresent(inv -> redis.opsForValue().set("stock:" + SEEDED_SKU, String.valueOf(inv.getStock())));
    }

    /** Redis Lua atomic decrement — one round-trip, no check-then-set race. */
    public long decrementRedis(String sku, long qty) {
        Long result = redis.execute(DECREMENT_LUA, List.of("stock:" + sku), String.valueOf(qty));
        if (result == null || result == -1L) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "unknown_sku");
        }
        if (result == -2L) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "insufficient_stock");
        }
        return result;
    }

    /** Pessimistic write lock decrement inside a transaction. */
    @Transactional
    public long decrementDb(String sku, int qty) {
        InventoryEntity inv = inventories.findBySkuForUpdate(sku)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "unknown_sku"));
        if (inv.getStock() < qty) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "insufficient_stock");
        }
        inv.setStock(inv.getStock() - qty);
        inventories.save(inv);
        // Append the audit row in the same transaction.
        ledgers.save(new InventoryLedgerEntity(sku, -qty, inv.getStock()));
        return inv.getStock();
    }

    public StockSnapshot readStock(String sku) {
        InventoryEntity inv = inventories.findById(sku)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "unknown_sku"));
        String redisRaw = redis.opsForValue().get("stock:" + sku);
        long redisStock = redisRaw == null ? -1 : Long.parseLong(redisRaw);
        List<LedgerLine> ledger = ledgers.findBySkuOrderByIdAsc(sku).stream()
                .map(row -> new LedgerLine(row.getDelta(), row.getRemaining()))
                .toList();
        return new StockSnapshot(sku, inv.getStock(), redisStock, ledger);
    }

    public record LedgerLine(int delta, int remaining) {
    }

    public record StockSnapshot(String sku, long dbStock, long redisStock, List<LedgerLine> ledger) {
    }
}

File: 0-high-concurrency-inventory-management/0-typescript/src/inventory/inventory.service.ts
import {
    ConflictException,
    Inject,
    Injectable,
    NotFoundException,
} from "@nestjs/common"
import { InjectDataSource } from "@nestjs/typeorm"
import { DataSource } from "typeorm"
import Redis from "ioredis"
import { REDIS_CLIENT } from "./redis.provider"
import { InventoryEntity, InventoryLedgerEntity } from "./inventory.entities"

/**
 * Lua runs atomically on Redis's single-threaded core: the check and the
 * decrement happen as one indivisible unit, so no other client can interleave.
 */
const DECREMENT_LUA = `
local stock = tonumber(redis.call('GET', KEYS[1]))
if stock == nil then return -1 end
if stock < tonumber(ARGV[1]) then return -2 end
return redis.call('DECRBY', KEYS[1], ARGV[1])
`

@Injectable()
export class InventoryService {
    public constructor(
        @Inject(REDIS_CLIENT) private readonly redis: Redis,
        @InjectDataSource() private readonly dataSource: DataSource,
    ) {}

    /** Redis Lua atomic decrement — one round-trip, no check-then-set race. */
    public async decrementRedis(sku: string, quantity: number): Promise<number> {
        const result = Number(
            await this.redis.eval(
                DECREMENT_LUA,
                1,
                `stock:${sku}`,
                String(quantity),
            ),
        )
        if (result === -1) throw new NotFoundException(`Unknown SKU: ${sku}`)
        if (result === -2) throw new ConflictException("insufficient_stock")
        return result // remaining stock after the atomic decrement
    }

    /** Pessimistic write lock decrement — serialize writers on the authoritative row. */
    public async decrementDb(sku: string, quantity: number): Promise<number> {
        return this.dataSource.transaction(async (manager) => {
            const inv = await manager.findOne(InventoryEntity, {
                where: { sku },
                lock: { mode: "pessimistic_write" },
            })
            if (!inv) throw new NotFoundException(`Unknown SKU: ${sku}`)
            if (inv.stock < quantity)
                throw new ConflictException("insufficient_stock")

            inv.stock -= quantity
            await manager.save(inv)

            // Append the audit row in the SAME transaction so stock and ledger
            // commit or roll back together.
            await manager.insert(InventoryLedgerEntity, {
                sku,
                delta: -quantity,
                remaining: inv.stock,
            })
            return inv.stock
        })
    }

    /** Read the durable DB stock, the hot Redis counter, and the ledger. */
    public async readStock(sku: string): Promise<{
        sku: string
        dbStock: number
        redisStock: number
        ledger: { delta: number; remaining: number }[]
    }> {
        const inv = await this.dataSource
            .getRepository(InventoryEntity)
            .findOne({ where: { sku } })
        if (!inv) throw new NotFoundException(`Unknown SKU: ${sku}`)

        const redisRaw = await this.redis.get(`stock:${sku}`)
        const ledgerRows = await this.dataSource
            .getRepository(InventoryLedgerEntity)
            .find({ where: { sku }, order: { id: "ASC" } })

        return {
            sku,
            dbStock: inv.stock,
            redisStock: redisRaw === null ? -1 : Number(redisRaw),
            ledger: ledgerRows.map((row) => ({
                delta: row.delta,
                remaining: row.remaining,
            })),
        }
    }

    /** Warm the Redis hot counter from the authoritative DB stock on startup. */
    public async warmCounter(sku: string): Promise<void> {
        const inv = await this.dataSource
            .getRepository(InventoryEntity)
            .findOne({ where: { sku } })
        if (inv) await this.redis.set(`stock:${sku}`, String(inv.stock))
    }
}
